---
title: "01-Assign_peak2gene"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```



```{r}
library(assertthat)
library(tidyverse)
library(glue)
library(ChIPseeker)
library(filesstrings)
library(rtracklayer)
library(GenomicFeatures)
# BiocManager::install("org.Hs.eg.db")
# BiocManager::install("org.Mm.eg.db")
```


#### Setup directories ####
```{r}
# Setup directory
print(paste('getwd():', getwd()))
getwd()


# Read directory of RNAseq and ATACseq logFC datasets
DGE_DPA_raw_data_dir <- "../data/DGE_DPA_raw_data"

# Read directory of Gencode genomes
gencode_dir <- "../data/gencodegenes"

# Create directory for ATAC peak assigned to gene datasets
peakAnno_df_dir <- "../results/peakAnno_df"
create_dir(peakAnno_df_dir)

# Experiment metadata
metadata_df_path <-  '../data/experiment_metadata_df.csv'
metadata_df <- read.table(metadata_df_path, header=TRUE, sep =',', check.names = F)
metadata_df <-metadata_df[!is.na(metadata_df$Contrast),]
```





#### Read RNA DGE logFC ####
```{r}
read_RNA_DGE_logFC <- function(experiment_id){
  RNAseq_file <- metadata_df[metadata_df$Id == experiment_id, ]$RNAseq_DGE
  RNAseq_path <- glue('./{DGE_DPA_raw_data_dir}/{RNAseq_file}')
  print(paste('Assert that', RNAseq_path, 'exists:', assert_that(file.exists(RNAseq_path))))
  RNAseq <- read.table(RNAseq_path, header=TRUE, fill = TRUE)
  return (RNAseq)
}
```


#### Read ATAC DPA ####
```{r}
read_ATAC_DPA_logFC <- function(experiment_id){
  ATACseq_file <- metadata_df[metadata_df$Id == experiment_id, ]$ATACseq_DPA
  ATACseq_path <- glue('./{DGE_DPA_raw_data_dir}/{ATACseq_file}')
  print(paste('Assert that', ATACseq_path, 'exists:', assert_that(file.exists(ATACseq_path))))
  ATACseq_bed <- read.table(ATACseq_path, header=TRUE)
  ATACseq_bed <- ATACseq_bed[, c("chr", "start", "end", "logFC", "t", "P_Value", "peak_id")] 
  ATACseq_bed_path <- glue("./{DGE_DPA_raw_data_dir}/bed_{ATACseq_file}")
  write.table(ATACseq_bed, ATACseq_bed_path, row.names = F, col.names = T, sep='\t', quote=F) # save bed file i
  return(ATACseq_bed_path)
}
```


#### Subset expressed genes from TxDb ####
1) get list of gene from RNAseq
2) filter TxDb.Hsapiens.UCSC.hg38.knownGene from 1
3) convert filtered TxDb dataframe to TxDb object
```{r}
subset_genome <- function(experiment_id, RNAseq){
  
  print(paste('Subset expressed genes from TxDb from', metadata_df[metadata_df$Id == experiment_id, ]$Genome, 'genome'))
  
  # human hg38 transcripts from gencode
  # downloaded from https://www.gencodegenes.org/human/release_41.html
  if (metadata_df[metadata_df$Id == experiment_id, ]$Genome == 'human'){
    mygtf <- mygtf_human }
  
  if (metadata_df[metadata_df$Id == experiment_id, ]$Genome == 'mouse'){  
    mygtf <- mygtf_mouse }
  
  if (metadata_df[metadata_df$Id == experiment_id, ]$filtered_txdb == 'YES'){
    print(paste('Subset genes from txdb that are expressed in RNA'))
    # Subset genes from txdb that are expressed in RNA
    subset.genes.ind = which(mygtf$gene_name %in% RNAseq$gene_symbol)
    # Subset rows
    mygtf.filt = mygtf[subset.genes.ind]
    # filter to protein_coding genes
    mygtf.filt2 = subset(mygtf.filt, gene_type == "protein_coding")
    # convert filtered gtf to TxDb
    myTxDb = makeTxDbFromGRanges(mygtf.filt2)
  }
  if (metadata_df[metadata_df$Id == experiment_id, ]$filtered_txdb == 'NO'){
    print(paste('Do NOT subset genes from txdb that are expressed in RNA'))
    myTxDb = makeTxDbFromGRanges(mygtf)}
  
  return(myTxDb)
}
```


#### Find gene associated with ATAC DPA peak id ####
```{r}
annotate_ATAC_peak <- function(ATACseq_bed_path, myTxDb, experiment_id){
  
  if (metadata_df[metadata_df$Id == experiment_id, ]$Genome == 'human'){annoDb="org.Hs.eg.db"}
  if (metadata_df[metadata_df$Id == experiment_id, ]$Genome == 'mouse'){annoDb="org.Mm.eg.db"}

  peakAnno <- annotatePeak(ATACseq_bed_path, 
                         tssRegion=c(-3000, 3000),
                         TxDb=myTxDb,  
                         level = "gene",
                         annoDb=annoDb)
  
  # format peanAnno as dataframe
  peakAnno_df <- as.data.frame(peakAnno)
  print(paste('Find gene associated with ATAC DPA peak id - ChipSeeker, # peaks:', dim(peakAnno_df)[1]))
  
  peakAnno_df$ENSEMBL <- gsub("\\..*","",peakAnno_df$geneId)
  
  # Subset ATAC peaks with distToTSS within 1e6
  peakAnno_df <- peakAnno_df[peakAnno_df$distanceToTSS < 1e6, ] 
  
  # Save merged_ATAC_RNA_logFC_df
  peakAnno_df_name <- paste0('peakAnno_', metadata_df[metadata_df$Id == experiment_id, ]$Contrast, ".csv")
  peakAnno_df_path <- paste0(glue("{peakAnno_df_dir}/{peakAnno_df_name}"))
  print(paste('Save peakAnno_df as:', peakAnno_df_path))
  write.csv(peakAnno_df, peakAnno_df_path, row.names = FALSE)
  
  ##### Save file name in metadata_df #####
  metadata_df <- read.table(metadata_df_path, header=TRUE, sep =',')
  metadata_df[metadata_df$Id == experiment_id,]$peakAnno_df <- peakAnno_df_name
  write.csv(metadata_df, metadata_df_path, row.names = FALSE)
}
```




#### Run all ####
```{r}
mygtf_human <- rtracklayer::import(glue("{gencode_dir}/hg38.gencode.v41.annotation.gtf.gz"))
mygtf_mouse <- rtracklayer::import(glue("{gencode_dir}/m38.gencode.vM1.annotation.gtf.gz"))
```


```{r}
for (experiment_id in metadata_df$Id){
  print(paste('experiment_id:', experiment_id))
  
  start_time <- Sys.time()
  RNAseq <- read_RNA_DGE_logFC(experiment_id)
  ATACseq_bed_path <- read_ATAC_DPA_logFC(experiment_id)
  myTxDb <- subset_genome(experiment_id, RNAseq)
  peakAnno_df <- annotate_ATAC_peak(ATACseq_bed_path, myTxDb,  experiment_id)
  end_time <- Sys.time()
  print(end_time - start_time)

}
```






